sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/partitioning/KeyGroupedPartitioning.java

```diff
@@ -18,27 +18,37 @@
 package org.apache.spark.sql.connector.read.partitioning;
 
 import org.apache.spark.annotation.Evolving;
-import org.apache.spark.sql.connector.read.PartitionReader;
+import org.apache.spark.sql.connector.expressions.Expression;
 
 /**
- * An interface to represent data distribution requirement, which specifies how the records should
- * be distributed among the data partitions (one {@link PartitionReader} outputs data for one
- * partition).
+ * Represents a partitioning where rows are split across partitions based on the expressions
+ * returned by {@link KeyGroupedPartitioning#clustering}.
  * <p>
- * Note that this interface has nothing to do with the data ordering inside one
- * partition(the output records of a single {@link PartitionReader}).
- * <p>
- * The instance of this interface is created and provided by Spark, then consumed by
- * {@link Partitioning#satisfy(Distribution)}. This means data source developers don't need to
- * implement this interface, but need to catch as more concrete implementations of this interface
- * as possible in {@link Partitioning#satisfy(Distribution)}.
- * <p>
- * Concrete implementations until now:
- * <ul>
- *   <li>{@link ClusteredDistribution}</li>
- * </ul>
+ * Data source implementations should make sure that all rows where
+ * {@link KeyGroupedPartitioning#clustering} evaluate to the same value should be in the same
+ * partition.
```

Comment:
```suggestion
 * partition, and in different partitions if the evaluated values are differnet.
```

Reply:
Actually, rows with the same value can also be in different partitions, because each partition can report its value and Spark will group partitions.

Reply:
I think we need to spend more time making this doc clear. `KeyGroupedPartitioning` does mean that, logiclally each output partition should only contains rows with the same value, and different partitions have different values. But physically the data source can put rows with the same value in different partitions, and Spark will group them later.

Reply:
Oops you are right. I forgot to update this. Will add.